hermes/probe/journal/journal.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	return nil
+}
+
+// reader returns an io reader that in turn returns the file contents.
+// Returns:
+// - reader: returns an io reader that in turn returns the file contents.
+func (j *journalFile) reader() io.Reader {
+	return bytes.NewReader(j.contents)
+}
+
+// unmarshalContents unmarshals the contents of the journal file and returns
+// the completed proto struct.
+// Returns:
+// 		- journal: proto struct containing the unmarshalled contents of the journal
+//		- err: error if once occurs during the unmarshalling process.
+func (j *journalFile) unmarshalContents() (*pb.StateJournal, *probe.ProbeError) {
+	journal := &pb.StateJournal{}
+	if err := proto.Unmarshal(j.contents, journal); err != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("verifyNilFile(): unable to unmarshal nil file proto: %w", err))
+	}
+	return journal, nil
+}
+
+// CheckJournal checks the existence and validity of the Journal file on the target
+// storage system. It also reads the intent from the Journal file to identify if
+// Hermes failed to complete the last operation, and completes it if necessary.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func CheckJournal(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	file, err := findJournal(ctx, target, client, logger)
+	if err != nil {
+		return err
+	}
+
+	if file.contents, err = readJournal(ctx, file.name, target, client, logger); err != nil {
+		return err
+	}
+
+	if err := file.checksumContents(); err != nil {
+		return err
+	}
+
+	journal, err := file.unmarshalContents()
+	if err != nil {
+		return err
+	}
+
+	if err := resolveIntent(ctx, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	if err = verifyJournal(ctx, file, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	logger.Debug("CheckJournal() completed successfully, nil file is valid.")
+	return nil
+}
+
+// findJournal checks that the journal file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- journalFile: a journalFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func findJournal(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) (*journalFile, *probe.ProbeError) {
```

Comment:
I would change this to return the filename only.

Reply:
I think recording the difference between the `ListFiles` query failing (an API call issue) and the journal file not being found (a file missing issue) is worthwhile here. If I return only the filename, then in the case of error, the return value would be `nil`.